docs/sphinx/core/config.rst

.. _marvin-config-info:

Configuration Class (marvin.config)
===================================
The Marvin :ref:`marvin-config-class` class controls the global configuration of Marvin.  It has
several attributes that control how you interact with MaNGA data.

* **mode**:
    The mode you want to run Marvin in. Mode can be either 'local', 'remote', or 'auto'. This determines how
    you interact with the MaNGA data.  Either through local FITS files on your own filesystem, or remotely using an
    API, if you do not have the data stored locally.  For more information, see :ref:`marvin-modes` below.

* **download**:
    Setting this attribute to True tells Marvin that whenever you don't have a file locally, to use
    sdss_access to download it via rsync.  See :ref:`marvin-download-objects` below.

* **release**:
    The Config class has a single attribute referencing the version of MaNGA data you are working with.
    **release** refers to either MaNGA Product Launches (MPLs), e.g. "MPL-5", or Data Releases (DRs), e.g. "DR13".  See :ref:`marvin-set-versions` below, to see how to set the version of the data you want.

* **use_sentry**:
    Marvin uses `Sentry <https://sentry.io>`_ to handle real-time logging of errors.  This setting toggles this feature.  The default value is set to **True**.  Set to **False** to disable.

* **add_github_message**:
    Marvin appends a message to every error instructing you on how to submit a new Github Issue regarding the error you just experienced.  If you wish to disable this message, set this value to **False**.  The default value is **True**.

* **db**:
    This attribute lets Marvin know if you have a database that it can be connected to.  If you have no database, this
    attribute will be set to None.  This attribute is set automatically and **you do not have to do anything with this attribute**.

* **sasurl**:
    This attribute tells Marvin what the SAS base url is when accessing the Marvin API. This attribute
    is set automatically and **you do not have to do anything with this attribute**.

* **urlmap**:
    This attribute is a lookup dictionary for all of the API routes that Marvin uses internally.
    This attribute is set automatically.  If you not using the Marvin API :ref:`marvin-api-interaction` class directly,
    then **you do not have to do anything with this attribute**.

* **access**:
    This attribute informs Marvin of the type of access it has.  The allowed values are either "public" or "collab", for public and collaborationa access, respectively.  Public access provides access only to MaNGA public Data Releases, while collaboration access provides access to all MaNGA data release, DRs and MPLs.  The default value is **public**.

* **login**:
    This method

.. _marvin-modes:

Data Access Modes
=================
* **Local mode** - Use this mode to deal with local FITS files either in your local SAS or through explicit file locations.
* **Remote mode** - Use this mode to deal with data remotely.  The data is retrieved from Utah using the API.  Depending on your use,
  it may be returned as a JSON object or used to complete the Tool function you are using.
* **Auto mode** - Use this mode to have Marvin attempt to automatically handle the modes.  Marvin starts in this mode. It is recommended to leave Marvin in this mode, and let him handle your data access.
* See :ref:`marvin-dam` for more detailed information.

.. _marvin-set-versions:

Setting Versions
================
You can globally control the version of the data you want to access using several convienence methods in the :ref:`marvin-config-class` class. Setting the release will also internally set up the appropriate DRP and DAP versions.

:py:meth:`~marvin.Config.setRelease`: main method to set the release version (e.g. MPL-5 or DR13).  Accepts either MPL or DR strings.

::

    from marvin import config
    config.setRelease('MPL-5')

You can also individually set MPLs or DRs separately with similar methods.

:py:meth:`~marvin.Config.setMPL`: set the version using the MPL designation (e.g. MPL-4)

::

    from marvin import config
    config.setMPL('MPL-4') # sets the global version to MPL-4

:py:meth:`~marvin.Config.setDR`: set the version using the DR designation (e.g. DR13)

::

    from marvin import config
    config.setDR('DR15') # sets the global version to DR15

.. _marvin-access:

Public vs Collab Access
=======================

There are two access modes in Marvin, one for the public, and one for members of the SDSS collaboration.  The default access mode in Marvin is **public**.  Public access only provides access to MaNGA released in public Data Releases (DRs).

::

    # the default release with public access is the latest DR
    from marvin import config
    INFO: No release version set. Setting default to DR15

    # the data release is DR15
    config.release
    'DR15'

If you are a member of the SDSS collaboration and have properly set up your netrc authentication, you can switch access to **collab** and get access to MaNGA collaboration data, MPLs, as well as DRs.

::

    # switch to collaboration access
    from marvin import config
    config.access = 'collab'

    # switch to an MPL
    config.setRelease('MPL-7')

.. _marvin-api-login:

Logging In
==========

The Marvin API requires authentication with a token for access.  To receive a token, use the ``login`` method
::

    # to receive an API token
    config.login()

    # check you have a token
    config.token

A valid token lasts for 300 days.  When your token expires, you will need to login again to receive a new token.
::

    # receive a fresh token
    config.login(refresh=True)

Note, by default your token will disappear upon exiting your iPython terminal session.  You will need to login again within a new session.  To preserve your token between iPython sessions, copy your token into the **use_token** attribute of your custom Marvin config file.

Attempting to use Marvin with invalid credentials or an invalid token will produce the following error.
::

    # access a cube remotely without proper credentials
    cube = Cube('8485-1901', mode='remote')

    MarvinError: found a problem when checking if remote cube exists: API Authentication Error: Token has expired. Please check your token or login again for a fresh one.


.. _marvin_custom_yaml:

Marvin Custom Configuration File
================================

Most Marvin configuration options can be set using the Marvin ``config`` object within iPython.  You can also set configuration parameters using a custom YAML configuration file, ``marvin.yml``.  This file must be placed in the following directory in your HOME, ``~/.marvin/marvin.yml``.  Currently configurable options are:

* **check_access**:
    Set to **True** to have Marvin automatically check for proper netrc collaboration access and switch to **collab** access mode on startup.  Default is **False**.

* **use_sentry**:
    Set to **False** to disable Sentry error logging in Marvin.  Default is **True**.

* **add_github_message**:
    Set to **False** to disable the Github Issue message on all Marvin Errors.  Default is **True**.

* **use_token**:
    Set this value to your valid API token.  This ensures proper API authentication across iPython sessions.

* **default_release**:
    Set to to the release you want to use by default when importing Marvin. If set to **null**, uses the latest available version for your access mode. Default is **null**.


CHANGELOG.md
# Marvin's Change Log

## [2.1.5] - unreleased
### Added:
- Better BPT documentation, in particular in the `Modifying the plot` section.
- A hack function ``marvin.utils.plot.utils.bind_to_figure()`` that replicate the contents of a matplotlib axes in another figure.

### Changed:
- Issue #190: ``Maps.get_bpt()`` and ``marvin.utils.dap.bpt.bpt_kewley06()`` now also return a list of axes. Each axes contains a method pointing to the ``marvin.utils.plot.utils.bind_to_figure()`` function, for easily transfer the axes to a new figure.

### Fixed:


## [2.1.4] - 2017/08/02
### Added:
- Added new query_params object, for easier navigation of available query parameters.  Added new tests.
- Added a new guided query builder using Jquery Query Builder to the Search page
- Added a View Galaxies link on the web results to view postage stamps of the galaxies in the results
- Added Route Rate Limiting.  Adopts a limit of 200/min for all api routes and 60/minute for query api calls and web searches

### Changed:
- Changed call signature for :meth:`marvin.utils.plot.map.no_coverage_mask` (removed ``value`` arg because unused, added ``None`` as default value ``ivar`` (``None``), and re-ordered args and kwargs).
- Changed call signature for :meth:`marvin.utils.plot.map.bad_data_mask` (removed ``value`` arg because unused).
- Changed the Marvin web search page to use the new query_params and parameter grouping.  Removed the autocomplete input box.
- Updated the documentation on query and query_params.
- Modified Guided Search operator options to remove options that could not be parsed by SQLA boolean_search
- Refactored the web settings, route registration, extensions to enable extensibility
- Issue #282: Improvements to "Go to CAS" link.  Changed to Go To SkyServer and updated link to public up-to-date link

### Fixed:
- Issue #102: problem with urllib package when attempting to retrieve the Marvin URLMap
- Issue #93: safari browser does not play well with marvin
- Issue #155: Contrails in Web Map
- Issue #174: sdss_access may not be completely python 3 compatible
- Issue #196: Bin not loading from local sas
- Issue #207: Get Maps in MapSpecView of Galaxy page sometimes fails to return selected maps
- Issue #210: pip upgrade may not install new things as fresh install
- Issue #209: marvin version from pip install is incorrect
- Issue #268: Cube flux from file error
- Issue #85: Python does not start in Python 3
- Issue #273: ha.value bug
- Issue #277: Ticks for log normalized colorbar
- Issue #275: logger crashes on warning when other loggers try to log
- Issue #258: 422 Invalid Parameters
- Issue #271: Problem in dowloading image.
- Issue #97: sqlalchemy-boolean-search not found when installed from pip source
- Issue #227: Marvin installation in python 3.6 (update setuptools to 36)
- Issue #262: problem with marvin update
- Issue #270: BPT array sizing not compatible
- Issue #88: Deployment at Utah requires automatisation
- Issue #234: Add (and use) functions to the datamodel to determine plotting parameters
- Issue #278: marvin_test_if decorator breaks in python 2.7
- Issue #274: cube slicing to get a spaxel fails with maps error
- Issue #39: implement more complete testing framework
- Issue #242: Result object representation error with 0 query results
- Issue #159: Marvin issues multiple warnings in PY3
- Issue #149: Improve integrated flux maps display in web

## [2.1.3] - 2017/05/18
### Added:
- Issue #204: added elpetro_absmag colours to mangaSampleDB models.
- Issue #253: Plotting tutorial.
- Issue #223: Easy multi-panel map plotting (with correctly placed colorbars).
- Issue #232 and Issue #251: Uses matplotlib style sheets context managers for plotting (map, spectrum, and BPT) and restores previous defaults before methods finish.
- Issue #189: Map plotting accepts user-defined value, ivar, and/or mask (including BPT masks).
- Issue #252: Quantile clipping for properties other than velocity, sigma, or flux in web.
- Added `utils.plot.map` doc page.
- Added `tools.map` doc page.

### Changed:
- Issue #243: inverted `__getitem__` behaviour for Cube/Maps/ModelCube and fixed tests.
- Modified Flask Profiler File to always point to $MARVIN_DIR/flask_profiler.sql
- Issue #241: Moved map plotting methods from tools/map to utils/plot/map
- Issue #229 and Issue #231: Switch to new gray/hatching scheme (in tools and web):
  - gray: spaxels with NOCOV.
  - hatched: spaxels with bad data (UNRELIABLE and DONOTUSE) or S/N below some minimum value.
  - colored: good data.
- Issue #238: Move plot defaults to datamodel (i.e., bitmasks, colormaps, percentile clips, symmetric, minimum SNR).
- Issue #206: SNR minimum to None (effectively 0) for velocity maps so that they aren't hatched near the zero velocity contour.
- Simplified default colormap name to "linearlab."
- Decreased map plot title font size in web so that it does not run onto second line and overlap plot.

### Fixed:
- Interactive prompt for username in sdss_access now works for Python 3.
- Fixed #195: The data file for the default colormap for `Map.plot()` ("linear_Lab") is now included in pip version of Marvin and does not throw invalid `FileNotFoundError` if the data file is missing.
- Fixed #143: prevents access mode to go in to remote if filename is present.
- Fixed #213: shortcuts are now only applied on full words, to avoid blind replacements.
- Fixed #206: no longer masks spaxels close to zero velocity contour in web and tools map plots
- Fixed #229: corrects web bitmask parsing for map plots
- Fixed #231: hatch regions within IFU but without data in map plots
- Fixed #255: Lean tutorial code cells did not work with the ipython directive, so they now use the python directive.
- Highcharts draggable legend cdn.

### Removed:
- Issue #232 and Issue #251: Automatic setting of matplotlib style sheets via seaborn import or `plt.style.use()`.


## [2.1.2] - 2017/03/17
### Added:

python/marvin/web/settings.py
# !usr/bin/env python2
# -*- coding: utf-8 -*-
#
# Licensed under a 3-clause BSD license.
#
# @Author: Brian Cherinka
# @Date:   2017-06-28 15:32:49
# @Last modified by:   Brian Cherinka
# @Last Modified time: 2018-11-17 13:15:31

from __future__ import print_function, division, absolute_import
import os
import datetime


class Config(object):
    SECRET_KEY = os.environ.get('MARVIN_SECRET', 'secret-key')
    FLASK_SECRET = SECRET_KEY
    APP_DIR = os.path.abspath(os.path.dirname(__file__))  # This directory
    APP_BASE = os.environ.get('MARVIN_BASE', 'marvin')
    projroot = os.path.abspath(os.path.join(APP_DIR, os.pardir, os.pardir, os.pardir))
    PROJECT_ROOT = os.environ.get('MARVIN_DIR', projroot)
    BCRYPT_LOG_ROUNDS = 13
    ASSETS_DEBUG = False
    DEBUG_TB_ENABLED = False  # Disable Debug toolbar
    DEBUG_TB_INTERCEPT_REDIRECTS = False
    CACHE_TYPE = 'memcached'  # Can be "memcached", "redis", etc.
    MAIL_SERVER = ''
    MAIL_PORT = 587
    MAIL_USE_SSL = False
    MAIL_USERNAME = ''
    MAIL_PASSWORD = ''
    MAIL_DEFAULT_SENDER = ''
    GOOGLE_ANALYTICS = ''
    LOG_SQL_QUERIES = True
    FEATURE_FLAGS_NEW = True
    UPLOAD_FOLDER = os.environ.get("MARVIN_DATA_DIR", '/tmp/')
    LIB_PATH = os.path.join(APP_DIR, 'lib')
    ALLOWED_EXTENSIONS = set(['txt', 'csv'])
    MAX_CONTENT_LENGTH = 50 * 1024 * 1024
    WTF_CSRF_ENABLED = True
    WTF_CSRF_SECRET_KEY = 'wtfsecretkey'
    USE_PROFILER = True  # Use the Flask Profiler Extension
    USE_SENTRY = False  # Turn off Sentry error logging
    FLASK_PROFILER = {
        "enabled": True,
        "storage": {
            "engine": "sqlite",
            "FILE": os.path.join(PROJECT_ROOT, 'flask_profiler.sql')
        },
        'endpointRoot': '{0}/profiler'.format(APP_BASE),
        "basicAuth": {
            "enabled": False
        },
        "ignore": [
            "/marvin/jsglue.js",
            "/marvin/static/.*",
            "/marvin/lib/.*",
            "/marvin/getgalidlist/"
        ]
    }
    # RATELIMIT_DEFAULT = '10/hour;100/day;2000 per year'
    RATELIMIT_STRATEGY = 'fixed-window-elastic-expiry'
    RATELIMIT_ENABLED = True


class ProdConfig(Config):
    """Production configuration."""
    ENV = 'prod'
    DEBUG = False
    SQLALCHEMY_DATABASE_URI = 'postgresql://localhost/example'
    DEBUG_TB_ENABLED = False  # Disable Debug toolbar
    USE_X_SENDFILE = True
    USE_SENTRY = True
    SENTRY_DSN = os.environ.get('SENTRY_DSN', None)
    PERMANENT_SESSION_LIFETIME = datetime.timedelta(1)
    JWT_ACCESS_TOKEN_EXPIRES = datetime.timedelta(300)
    REMEMBER_COOKIE_DOMAIN = '.sdss.org'


class DevConfig(Config):
    """Development configuration."""
    ENV = 'dev'
    DEBUG = True
    DB_NAME = 'dev.db'
    # Put the db file in project root
    DB_PATH = os.path.join(Config.PROJECT_ROOT, DB_NAME)
    SQLALCHEMY_DATABASE_URI = 'sqlite:///{0}'.format(DB_PATH)
    DEBUG_TB_ENABLED = True
    ASSETS_DEBUG = True  # Don't bundle/minify static assets
    CACHE_TYPE = 'simple'  # Can be "memcached", "redis", etc.
    RATELIMIT_ENABLED = False


class TestConfig(Config):
    TESTING = True
    DEBUG = True
    SQLALCHEMY_DATABASE_URI = 'sqlite://'
    BCRYPT_LOG_ROUNDS = 1  # For faster tests
    WTF_CSRF_ENABLED = False  # Allows form testing
    PRESERVE_CONTEXT_ON_EXCEPTION = False
    USE_PROFILER = False  # Turn off the Flask Profiler extension
    RATELIMIT_ENABLED = False  # Turn off the Flask Rate Limiter
    #os.environ['PUBLIC_SERVER'] = 'True' # this breaks the debug server


class CustomConfig(object):
    ''' Project specific configuration.  Always gets appended to an above Config class '''
    os.environ['SAS_REDUX'] = 'sas/mangawork/manga/spectro/redux'
    os.environ['SAS_ANALYSIS'] = 'sas/mangawork/manga/spectro/analysis'
    os.environ['SAS_SANDBOX'] = 'sas/mangawork/manga/sandbox'
    release = os.environ.get('MARVIN_RELEASE', 'mangawork')
    os.environ['SAS_PREFIX'] = 'marvin' if release == 'mangawork' else 'dr15/marvin'
